### PR TITLE
Delete React Call/Return experiment

### PR DESCRIPTION
**Pull request by gaearon**
We're not using this anywhere and it was always marked as completely unsafe so nobody should rely on this. It's taking up bundle size but per conversation with @sebmarkbage this model is insufficient for the layout use case he was interested in. Let's get rid of this and free up some space for actual features we're shipping, and come back to this when we have a better API in mind.

### CODE REVIEW

File: packages/shared/ReactTypeOfWork.js [hunk 1]
```diff
@@ -33,9 +33,9 @@ export const HostRoot = 3; // Root of a host tree. Could be nested inside anothe
 export const HostPortal = 4; // A subtree. Could be an entry point to a different renderer.
 export const HostComponent = 5;
 export const HostText = 6;
-export const CallComponent = 7;
-export const CallHandlerPhase = 8;
-export const ReturnComponent = 9;
+export const CallComponent_UNUSED = 7;
```
Comment (acdlite) [inline packages/shared/ReactTypeOfWork.js]: Why are we leaving these in?

Comment (gaearon) [reply packages/shared/ReactTypeOfWork.js]: Might rename/reuse them for next types we add?

**Comment by gaearon** [comment]
Just so we can use them for next types instead of leaving a hole. Since they're effectively "unclaimed".

**Comment by gaearon** [comment]
The symbols are also ordered (`0xeac7` -> `0xeac8` -> ...). So I'd prefer to keep them and then "fill them in" as necessary. I really don't care though. :-)

File: packages/shared/ReactTypeOfWork.js (hunk 1 above)
Comment (sebmarkbage) [reply packages/shared/ReactTypeOfWork.js]: Does anything rely on the type of work number? Just delete and collapse?

**Comment by react-linter** [comment]
ReactDOM: size: :small_red_triangle:+1.3%, gzip: :small_red_triangle:+0.9%


  <details>
  <summary>Details of bundled changes.</summary>

  <p>Comparing: d758960116b3ea3a3bd3dfe2f297d5e680fe9ff5...b6857a910d4d1e65d276429a9a4367acd7796570</p>


  
## react-dom
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-dom.development.js | -0.7% | -0.6% | 646.72 KB | 641.95 KB | 148.4 KB | 147.53 KB | UMD_DEV
**react-dom.production.min.js** | **:small_red_triangle:+1.3%** | **:small_red_triangle:+0.9%** | **101.12 KB** | **102.44 KB** | **32.36 KB** | **32.63 KB** | **UMD_PROD**
react-dom.development.js | -0.8% | -0.6% | 631.08 KB | 626.32 KB | 144.33 KB | 143.4 KB | NODE_DEV
**react-dom.production.min.js** | **:small_red_triangle:+1.3%** | **:small_red_triangle:+0.9%** | **99.53 KB** | **100.85 KB** | **31.45 KB** | **31.74 KB** | **NODE_PROD**
react-dom-test-utils.development.js | -9.6% | -5.5% | 45.03 KB | 40.71 KB | 12.36 KB | 11.68 KB | UMD_DEV
**react-dom-test-utils.production.min.js** | **:small_red_triangle:+1.7%** | **:small_red_triangle:+2.7%** | **10.42 KB** | **10.6 KB** | **3.84 KB** | **3.94 KB** | **UMD_PROD**
react-dom-test-utils.development.js | -10.8% | -6.2% | 39.89 KB | 35.57 KB | 10.93 KB | 10.26 KB | NODE_DEV
**react-dom-test-utils.production.min.js** | **:small_red_triangle:+1.8%** | **:small_red_triangle:+2.9%** | **9.69 KB** | **9.87 KB** | **3.62 KB** | **3.73 KB** | **NODE_PROD**
react-dom-unstable-native-dependencies.development.js | -0.4% | -0.9% | 61.45 KB | 61.18 KB | 16.17 KB | 16.02 KB | UMD_DEV
**react-dom-unstable-native-dependencies.production.min.js** | **-0.4%** | **-0.2%** | **11.36 KB** | **11.32 KB** | **3.92 KB** | **3.91 KB** | **UMD_PROD**
react-dom-unstable-native-dependencies.development.js | -0.5% | -0.9% | 57.11 KB | 56.84 KB | 14.9 KB | 14.77 KB | NODE_DEV
**react-dom-unstable-native-dependencies.production.min.js** | **-0.4%** | **:small_red_triangle:+0.1%** | **10.71 KB** | **10.66 KB** | **3.67 KB** | **3.67 KB** | **NODE_PROD**
react-dom-server.browser.development.js | -0.4% | -0.2% | 101.97 KB | 101.58 KB | 26.6 KB | 26.54 KB | UMD_DEV
**react-dom-server.browser.production.min.js** | **:small_red_triangle:+0.3%** | **:small_red_triangle:+0.2%** | **15.17 KB** | **15.23 KB** | **5.81 KB** | **5.82 KB** | **UMD_PROD**
react-dom-server.browser.development.js | -0.4% | -0.2% | 91.27 KB | 90.88 KB | 24.34 KB | 24.28 KB | NODE_DEV
**react-dom-server.browser.production.min.js** | **:small_red_triangle:+0.4%** | **:small_red_triangle:+0.2%** | **14.54 KB** | **14.59 KB** | **5.55 KB** | **5.56 KB** | **NODE_PROD**
react-dom-server.node.development.js | -0.4% | -0.2% | 93.2 KB | 92.8 KB | 24.88 KB | 24.83 KB | NODE_DEV
**react-dom-server.node.production.min.js** | **:small_red_triangle:+0.3%** | **:small_red_triangle:+0.2%** | **15.34 KB** | **15.39 KB** | **5.84 KB** | **5.86 KB** | **NODE_PROD**
ReactDOM-dev.js | -0.8% | -0.7% | 656.54 KB | 651.53 KB | 147.3 KB | 146.28 KB | FB_WWW_DEV
**ReactDOM-prod.js** | **:small_red_triangle:+0.5%** | **:small_red_triangle:+0.4%** | **302.4 KB** | **304.01 KB** | **54.91 KB** | **55.11 KB** | **FB_WWW_PROD**
ReactTestUtils-dev.js | -10.7% | -6.1% | 41.29 KB | 36.85 KB | 11.13 KB | 10.45 KB | FB_WWW_DEV
ReactDOMUnstableNativeDependencies-dev.js | -0.4% | -0.9% | 57.32 KB | 57.09 KB | 14.69 KB | 14.56 KB | FB_WWW_DEV
**ReactDOMUnstableNativeDependencies-prod.js** | **-0.0%** | **0.0%** | **26.35 KB** | **26.34 KB** | **5.37 KB** | **5.37 KB** | **FB_WWW_PROD**
ReactDOMServer-dev.js | -0.6% | -0.3% | 94.98 KB | 94.43 KB | 24.17 KB | 24.1 KB | FB_WWW_DEV
**ReactDOMServer-prod.js** | **:small_red_triangle:+0.2%** | **:small_red_triangle:+0.1%** | **31.65 KB** | **31.71 KB** | **7.77 KB** | **7.78 KB** | **FB_WWW_PROD**

## react-art
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-art.development.js | -0.2% | -0.3% | 445.75 KB | 444.95 KB | 96.88 KB | 96.57 KB | UMD_DEV
**react-art.production.min.js** | **:small_red_triangle:+0.1%** | **-0.2%** | **92.64 KB** | **92.75 KB** | **28.24 KB** | **28.18 KB** | **UMD_PROD**
react-art.development.js | -0.2% | -0.2% | 371.58 KB | 370.78 KB | 78.01 KB | 77.89 KB | NODE_DEV
**react-art.production.min.js** | **:small_red_triangle:+0.2%** | **-0.1%** | **57.15 KB** | **57.26 KB** | **17.5 KB** | **17.48 KB** | **NODE_PROD**
ReactART-dev.js | -0.2% | -0.2% | 380.59 KB | 379.81 KB | 77.75 KB | 77.59 KB | FB_WWW_DEV
**ReactART-prod.js** | **0.0%** | **-0.0%** | **186.66 KB** | **186.68 KB** | **30.64 KB** | **30.63 KB** | **FB_WWW_PROD**

## react-test-renderer
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-test-renderer.development.js | -0.1% | -0.1% | 376.54 KB | 376.07 KB | 78.87 KB | 78.77 KB | UMD_DEV
**react-test-renderer.production.min.js** | **:small_red_triangle:+0.2%** | **0.0%** | **56.89 KB** | **57.01 KB** | **17.28 KB** | **17.28 KB** | **UMD_PROD**
react-test-renderer.development.js | -0.1% | -0.1% | 367.36 KB | 366.88 KB | 76.19 KB | 76.09 KB | NODE_DEV
**react-test-renderer.production.min.js** | **:small_red_triangle:+0.2%** | **-0.2%** | **56.12 KB** | **56.23 KB** | **16.99 KB** | **16.96 KB** | **NODE_PROD**
react-test-renderer-shallow.development.js | -0.5% | -0.4% | 24.5 KB | 24.38 KB | 6.53 KB | 6.5 KB | UMD_DEV
**react-test-renderer-shallow.production.min.js** | **:small_red_triangle:+0.1%** | **:small_red_triangle:+0.3%** | **7.2 KB** | **7.21 KB** | **2.35 KB** | **2.36 KB** | **UMD_PROD**
react-test-renderer-shallow.development.js | -1.8% | -1.5% | 14.54 KB | 14.27 KB | 3.63 KB | 3.58 KB | NODE_DEV
**react-test-renderer-shallow.production.min.js** | **-1.8%** | **-1.4%** | **7.28 KB** | **7.15 KB** | **2.38 KB** | **2.34 KB** | **NODE_PROD**
ReactTestRenderer-dev.js | -0.1% | -0.2% | 376.55 KB | 376.19 KB | 76.03 KB | 75.91 KB | FB_WWW_DEV
ReactShallowRenderer-dev.js | -1.8% | -1.4% | 15.28 KB | 15.01 KB | 3.75 KB | 3.7 KB | FB_WWW_DEV

## react-noop-renderer
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-noop-renderer.development.js | +2.7% | +8.6% | 17.87 KB | 18.35 KB | 4.69 KB | 5.1 KB | NODE_DEV
**react-noop-renderer.production.min.js** | **-3.8%** | **-2.2%** | **6.6 KB** | **6.35 KB** | **2.56 KB** | **2.5 KB** | **NODE_PROD**

## react-reconciler
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-reconciler.development.js | -0.2% | -0.1% | 347.66 KB | 346.9 KB | 71.59 KB | 71.49 KB | NODE_DEV
**react-reconciler.production.min.js** | **:small_red_triangle:+0.2%** | **0.0%** | **48.85 KB** | **48.97 KB** | **14.74 KB** | **14.74 KB** | **NODE_PROD**
react-reconciler-persistent.development.js | -0.2% | -0.1% | 346.86 KB | 346.11 KB | 71.29 KB | 71.18 KB | NODE_DEV
**react-reconciler-persistent.production.min.js** | **:small_red_triangle:+0.2%** | **0.0%** | **47.79 KB** | **47.91 KB** | **14.54 KB** | **14.54 KB** | **NODE_PROD**
react-reconciler-reflection.development.js | -2.3% | -1.6% | 11.69 KB | 11.42 KB | 3.57 KB | 3.52 KB | NODE_DEV
**react-reconciler-reflection.production.min.js** | **:small_red_triangle:+0.4%** | **:small_red_triangle:+0.8%** | **2.26 KB** | **2.27 KB** | **1 KB** | **1.01 KB** | **NODE_PROD**

## react-native-renderer
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
ReactNativeRenderer-dev.js | -0.2% | +0.1% | 488.52 KB | 487.38 KB | 103.84 KB | 103.91 KB | RN_FB_DEV
**ReactNativeRenderer-prod.js** | **-0.2%** | **-0.6%** | **236.55 KB** | **236.18 KB** | **39.56 KB** | **39.33 KB** | **RN_FB_PROD**
ReactNativeRenderer-dev.js | -0.2% | +0.1% | 488.18 KB | 487.04 KB | 103.75 KB | 103.83 KB | RN_OSS_DEV
**ReactNativeRenderer-prod.js** | **-0.2%** | **-0.6%** | **222.76 KB** | **222.4 KB** | **37.39 KB** | **37.15 KB** | **RN_OSS_PROD**
ReactFabric-dev.js | -0.4% | -0.6% | 471.79 KB | 469.78 KB | 100.08 KB | 99.51 KB | RN_FB_DEV
**ReactFabric-prod.js** | **-0.6%** | **-1.0%** | **209.23 KB** | **208 KB** | **34.86 KB** | **34.53 KB** | **RN_FB_PROD**
ReactFabric-dev.js | -0.4% | -0.6% | 471.83 KB | 469.82 KB | 100.09 KB | 99.53 KB | RN_OSS_DEV
**ReactFabric-prod.js** | **-0.6%** | **-1.0%** | **209.26 KB** | **208.04 KB** | **34.88 KB** | **34.55 KB** | **RN_OSS_PROD**

  </details>
  
<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

File: packages/shared/ReactTypeOfWork.js (hunk 1 above)
Comment (gaearon) [reply packages/shared/ReactTypeOfWork.js]: DevTools does.

**Comment by rtsao** [comment]
Out of curiosity, what were the limitations of this model in terms of the mentioned layout use case? I'd love to hear more about that.

I've been toying around with this experiment to try and implement [Subform](https://subformapp.com) layout, which so far has been promising.

**Comment by k15a** [comment]
Is there any way this functionality can be implemented in userspace with the current React version? Maybe with the new context API?

**Comment by gaearon** [comment]
No. (If it was possible we wouldn’t do it in the core.)

**Comment by k15a** [comment]
Appreciate your response. I thought a little bit more about call return because I think it would make compound components a lot nicer. Can you please explain what's the major difference between call return and the [CellMeasurer of react-virtulized](https://bvaughn.github.io/react-virtualized/#/components/CellMeasurer)? As far as I understand it call return could be implemented in a similar way that CellMeasurer is. What's the major difference here?

**Comment by gaearon** [comment]
I'm not familiar with CellMeasurer. Maybe it's that call/return doesn't actually need to measure DOM nodes even if it's used for layout? It's more like a function call: it lets a parent know something about what happens *inside* a child component (if it opts in) during the rendering pass.

**Comment by BenLorantfy** [comment]
@k15a I almost kinda re-created this using only React but I had to make an awkward use of `document.querySelectorAll`. (which means it only works for react-dom).

https://github.com/BenLorantfy/react-lift-props
https://github.com/BenLorantfy/react-lift-props/blob/master/src/index.tsx#L41
https://medium.com/@benlorantfy/introducing-react-lift-props-490ee7cd958

If there was a way to get the "index" of a component in React, then I could avoid `document.querySelectorAll`.

```
<App>
  <Stepper>
    <MyStep>
      <Step /> <--- this would be index 0
    </MyStep>
    <MySecondStep>
      <Step /> <--- this would be index 1
    </MySecondStep>
  </Stepper>
</App>

**Comment by k15a** [comment]
@BenLorantfy I got something similar to work with Context and Portals.
https://gist.github.com/k15a/e3dd7313b280af9a47a32845180cb8ef

**Comment by BenLorantfy** [comment]
Has there been any progress/ideas on a replacement for this API?

**Comment by nihgwu** [comment]
I think my package [create-slots](https://github.com/nihgwu/create-slots) solves the same problem with a simpler mental model

**Pull request merged**